### PR TITLE
Prunability

### PR DESCRIPTION
**Pull request by berkaysynnada**
## Which issue does this PR close?

Closes #.

## Rationale for this change

We can have more optimized join operations by pruning the tables considering the finalized rows, or even be able to work on streams with bounded memory. Knowing whether we can do that pruning is a critical step in this process. This PR introduces a function that takes the parameters of a join filter and the parameters of the tables, then tells which table/tables is/are prunable, and some utility functions that can be helpful for other scopes, too. 

The algorithm starts by assigning an order and table-side information to each leaf node in PhysicalExpr. Then, it reaches the root node by propagating this information in a post-order way. During this propagation, each operator has its own rules. There are also some transformations of columns, sort expressions, and equivalence properties before the algorithm starts to 
be compatible with the intermediate schema of the join.

Example case: 
Predicate expression = a+3 < b AND c-d >= 10
a: increasing column from left table
b: increasing column from right table
c: decreasing column from right table
d: decreasing column from left table
With this information, we expect the function to return that we can prune the right side.

The first step is assignment:
image

After the propagation, the tree becomes like that:
image

As seen from the figures, each operator has some rules about how to propagate its children states. There are 2 important points:
1) Although an expression like ((c-d) >= 10) would be an 'unordered - none' node, an equivalence expression like (c >= (10 + d)) would be an 'unordered - left' node. This happens because if a numeric operator (such as +, -) has children coming from different tables, we cannot determine which table is prunable. However, a comparison operator can figure out that according to the sort order of children. This issue will be solved with a new PR, collecting the columns of the same table on the same side of the comparison operators before the prunability check.
2) All expressions carry their table side as if they will be prunable after the evaluation. However, for this to be possible, a comparison operator is required along the path. There is a flag for that purpose, and if there is not any comparison operator in the overall expression, though the end state of the root says that it is prunable, we will not care it and declare the tables as not prunable.

## What changes are included in this PR?

1) `is_filter_expr_prunable()` function that determines prunable table sides
2) Some utility functions to reshape `SortOptions` and `EquivalenceProperties` to be used in join's intermediate schema
3) `get_ordering()` method to calculate order of `PhysicalExpr`'s
4) `SortProperties` struct to differentiate literals with unordered data while propagating order information

## Are these changes tested?

Yes, we have extensive test cases for `is_filter_expr_prunable()`, and also have unit tests for some functions that can be used in other scopes.

### CODE REVIEW

File: datafusion/core/src/physical_plan/joins/utils.rs [hunk 1]
```diff
@@ -1400,13 +1454,699 @@ impl BuildProbeJoinMetrics {
...
+    equal_properties: F,
+    ordering_equal_properties: F2,
+    left_indices: &[(usize, &ColumnIndex)],
+    right_indices: &[(usize, &ColumnIndex)],
+) -> Result<Transformed<ExprPrunability>> {
+    // if we can directly match a sort expr with the current node, we can set its state and return early.
+    // TODO: If there is a PhysicalExpr other than Column at the node (let's say a+b), and there is an
+    // ordering equivalence of it (let's say c+d), we actually can find it at this step.
+    if check_direct_matching(&mut node, left_sort_expr, right_sort_expr) {
+        return Ok(Transformed::Yes(node));
```
Comment (metesynnada) [inline datafusion/core/src/physical_plan/joins/utils.rs:1658]:
```suggestion
        return Ok(Transformed::No(node));
```

**Review by metesynnada — COMMENTED** [review]
Add dosctring to 

- `intermediate_schema_sort_expr`
- `collect_one_side_columns`
- `merge_equivalence_classes_for_intermediate_schema`
- `assign_column_ordering`

Add tests for

- `merge_equivalence_classes_for_intermediate_schema`

**Comment by ozankabak** [comment]
Please clear out the merge conflicts and get the change count to a manageable number, I will review then.

File: datafusion/core/src/physical_optimizer/pipeline_checker.rs [hunk 2]
```diff
@@ -154,21 +156,32 @@ pub fn check_finiteness_requirements(
...
-/// This function returns whether a given symmetric hash join is amenable to
-/// data pruning. For this to be possible, it needs to have a filter where
-/// all involved [`PhysicalExpr`]s, [`Operator`]s and data types support
-/// interval calculations.
-///
-/// [`PhysicalExpr`]: crate::physical_plan::PhysicalExpr
-/// [`Operator`]: datafusion_expr::Operator
-fn is_prunable(join: &SymmetricHashJoinExec) -> bool {
+/// This function returns whether a given symmetric hash join is amenable to data pruning.
+fn is_prunable(join: &SymmetricHashJoinExec, children_unbounded: &[bool]) -> bool {
```
Comment (ozankabak) [inline datafusion/core/src/physical_optimizer/pipeline_checker.rs:160]: I refactored the logic in this function, please verify that it is still correct.

File: datafusion/core/src/physical_plan/joins/utils.rs [hunk 3]
```diff
@@ -1413,804 +1457,2918 @@ impl BuildProbeJoinMetrics {
...
+}
 
-    #[test]
-    fn check_valid() -> Result<()> {
-        let left = vec![Column::new("a", 0), Column::new("b1", 1)];
-        let right = vec![Column::new("a", 0), Column::new("b2", 1)];
-        let on = &[(Column::new("a", 0), Column::new("a", 0))];
+/// Takes the original sort expression of a table, and returns the updated
+/// version of it according to the intermediate schema of the join operator.
+fn intermediate_schema_sort_expr(
```
Comment (ozankabak) [inline datafusion/core/src/physical_plan/joins/utils.rs:1557]: I got rid of the `unwrap` in this function, please verify that this version is correct too.

File: datafusion/core/src/physical_plan/joins/utils.rs [hunk 4]
```diff
@@ -1413,804 +1457,2918 @@ impl BuildProbeJoinMetrics {
...
+    Both,
+}
 
-        assert!(check(&left, &right, on).is_err());
+/// Calculates the a [`PhysicalExpr`] node's [`ExprPrunability`] from its children.
+/// The [`TableSide`] part is updated in this function's scope, while [`SortProperties`]
+/// part of the state is updated in trait implementations of [`PhysicalExpr`]. The only
+/// exception is [`Column`] implementation as it needs a special handling considering
+/// the equivalence properties.
+fn update_monotonicity<
```
Comment (ozankabak) [inline datafusion/core/src/physical_plan/joins/utils.rs:1656]: Please verify that my refactor of this function remains correct.

File: datafusion/core/src/physical_plan/joins/utils.rs [hunk 5]
```diff
@@ -1963,7 +1960,8 @@ fn add_new_equivalences(
     initial_eq
 }
 
-/// Updates the [`PhysicalSortExpr`]'s according to the new names and column indices of intermediate schema.
+/// Updates the [`PhysicalSortExpr`]s according to the new names and column
+/// indices of the intermediate schema.
 fn transform_orders(
```
Comment (ozankabak) [inline datafusion/core/src/physical_plan/joins/utils.rs:1965]: I removed the inner `return Ok(Transformed::No(expr));` statement, which seemed unnecessary. Please verify that this version is still correct.

**Review by ozankabak — COMMENTED** [review]
The logic looks good to me overall. I did some refactors to some of your functions, please verify that they are correct. Here is the remaining work:

(1) The following functions/structs are missing docstrings, let's add:
- `normalize_ordering_equivalence_classes`
- `add_ordering_head_class`
- `add_ordering_other_classes`
- `get_tableside_at_numeric`
- `get_tableside_at_gt_or_gteq`
- `get_tableside_at_and`
- `TableSide`

Some other functions have docstrings, but those docstrings do not explain arguments etc. Please standardize all docstrings so that they also explain the arguments and the return value in addition to talking about what the function does.

(2) Write a nice PR body that will serve as our internal documentation for this PR.

**Pull request merged**